packages/blade/src/components/Box/BaseBox/getBaseBoxStyles.ts

```diff
@@ -1,102 +1,239 @@
-  value,
-  theme,
-}: {
-  value: SpacingValue;
-  theme: Theme;
-}): string | undefined => {
-  if (value === 'auto') return 'auto';
-  /**
-   * NOTE: don't allow numbers once we have the scale for sizing
-   * DO NOT PUBLISH THIS COMPONENT PUBLICLY WITH THIS PIECE OF CODE
-   */
-  if (typeof value === 'number') {
-    return makeSpace(value);
+const getResponsiveValue = <T extends string | number | string[]>(
+  value: MakeValueResponsive<T> | undefined,
+  size: keyof Breakpoints = 'base',
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+): any | undefined => {
+  if (value === undefined || value === null) {
+    return undefined;
   }
-  return value ? makeSpace(getIn(theme, value)) : undefined;
+
+  if (typeof value === 'string' || typeof value === 'number' || Array.isArray(value)) {
+    if (size === 'base') {
+      return value;
+    }
+
+    // Plain values will already get added in base styles so we don't have to repeat them in media queries
+    return undefined;
+  }
+
+  if (isEmpty(value)) {
+    return undefined;
+  }
+
+  if (isReactNative()) {
+    // In React Native, we map the value `s` token on priority (since the breakpoint maps to mobiles in useBreakpoint hook).
+    // We further look into smaller sizes, then we check base size.
+    // Then we return the first non-undefined value in this priority
+    const priorityArray = [value.s, value.xs, value.base];
+    return priorityArray.find((val) => val !== undefined);
+  }
+
+  return value[size];
+};
+
+const getSpacingValue = (
+  spacingValue:
+    | MakeValueResponsive<SpacingValueType | ArrayOfMaxLength4<SpacingValueType>>
+    | undefined,
+  theme: Theme,
+  size?: keyof Breakpoints,
+): string | undefined => {
+  if (isEmpty(spacingValue)) {
+    return undefined;
+  }
+
+  const responsiveSpacingValue: SpacingValueType | SpacingValueType[] = getResponsiveValue(
+    spacingValue as MakeValueResponsive<string | string[]>,
+    size,
+  );
+
+  if (isEmpty(responsiveSpacingValue)) {
+    return undefined;
+  }
+
+  if (responsiveSpacingValue === 'auto') {
+    return responsiveSpacingValue;
+  }
+
+  if (Array.isArray(responsiveSpacingValue)) {
+    return responsiveSpacingValue.map((value) => getSpacingValue(value, theme)).join(' ');
+  }
+
+  if (typeof responsiveSpacingValue === 'string' && responsiveSpacingValue.startsWith('spacing.')) {
+    const spacingReturnValue = getIn(theme, responsiveSpacingValue);
+    return isEmpty(spacingReturnValue) ? makeSpace(spacingReturnValue) : undefined;
+  }
+
+  // pixel or with unit values
+  return responsiveSpacingValue;
+};
+
+const getBackgroundValue = (
+  backgroundColor: BaseBoxProps['backgroundColor'],
+  theme: Theme,
+  size?: keyof Breakpoints,
+): string => {
+  const responsiveBackgroundValue = getResponsiveValue(backgroundColor, size);
+  const tokenValue = getIn(theme, `colors.${responsiveBackgroundValue}`);
+  return tokenValue ?? responsiveBackgroundValue;
```

Comment:
https://github.com/razorpay/blade/pull/1009#discussion_r1119663353

Reply:
Oh yup I meant, `getIn` will return undefined when it doesn't find that value no? so that `tokenValue ?? responsiveBackgroundValue` check will ignore undefined and return `red`.

It is expected. In the comment above I mentioned that there are several instances in our current codebase where we are passing backgroundColor string (since that is what the earlier BaseBox supported). So it will be lots of unrelated refactors for this PR to go and fix those instances right now.

Reply:
We support this only on BaseBox. On Box, we have added validations.